### PR TITLE
✨Adding Scrums

### PR DESCRIPTION
**Pull request by sam0000-cmd**
Added Scrums

[Notion Task Link] https://www.notion.so/Scrums-ea9a43e7bc0548b9bc6dff7032da0081

#### Branches for testing

Frontend: N/A

Backend: 'scrums-yug'

Bot: N/A

### CODE REVIEW

File: devsnest/app/controllers/api/v1/scrums_controller.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      include JSONAPI::ActsAsResourceController
+      before_action :user_auth
+      before_action :authorize_get, only: %i[index]
+      before_action :authorize_update, only: %i[update delete]
+      
+
+      def context
+        {
+          user:  @current_user,
+          scrum:  Scrum.find_by(id: params[:id]),
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/scrums_controller.rb:13]: where are you using scrum

File: devsnest/app/controllers/api/v1/scrums_controller.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      end
+
+      def authorize_update
+        scrum = Scrum.find_by(id: params[:id])
+        group = Group.find_by(id: params[:group_id])
+        return true if @current_user.id == scrum.user_id || @current_user.id == group.owner_id|| @current_user.id == group.co_owner_id || @current_user.id == group.batch_leader_id ||@current_user.user_type == 'admin'
+        render_error('message':'You Cannot Update this Scrum.')
+      end
+
+      def create
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/scrums_controller.rb:32]: can we create this function in scrums resource

File: devsnest/app/controllers/api/v1/scrums_controller.rb (hunk 1 above)
Comment (sam0000-cmd) [reply devsnest/app/controllers/api/v1/scrums_controller.rb:13]: Removed Scrum from context

File: devsnest/app/controllers/api/v1/scrums_controller.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+      def authorize_get
+        group = Group.find_by(id: params[:group_id])
+        return render_forbidden unless group.check_auth(@current_user)
+      end
+
+      def authorize_update
+        scrum = Scrum.find_by(id: params[:id])
+        group = Group.find_by(id: params[:group_id])
+        return true if @current_user.id == scrum.user_id || Scrum.check_scrum_auth(@current_user.id,group)
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/scrums_controller.rb:24]: pass the user na?

File: devsnest/app/models/scrum.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class Scrum < ApplicationRecord
+    validates :user_id, uniqueness: { scope: :date}
+
+    def self.check_scrum_auth(user_id,group)
+        user = User.find_by(id:user_id)
```
Comment (vedansh) [inline devsnest/app/models/scrum.rb:7]: remove this line, pass the user

Comment (vedansh) [reply devsnest/app/models/scrum.rb:7]: This function has nothing to do with scrum, this should be in a group, we can mark it as an instance function of group and call it admin-level auth or something

File: devsnest/app/resources/api/v1/scrum_resource.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      attributes :user_id, :group_id, :attendance, :saw_last_lecture,
+                  :till_which_tha_you_are_done, :what_will_you_cover_today, :reason_for_backlog_if_any, :rate_yesterday_class, :date
+
+      def self.creatable_fields(context)
+        group = Group.find_by(id: context[:group_id])
+        user_id = context[:user].id
+        unless Scrum.check_scrum_auth(user_id,group)
+          super - %i[attendance]
+        else
+          super - %i[]
```
Comment (vedansh) [inline devsnest/app/resources/api/v1/scrum_resource.rb:15]: sirf super

Comment (vedansh) [reply devsnest/app/resources/api/v1/scrum_resource.rb:15]: user_id, group_id bhi humme add karni chahiye, create hote samay

Comment (vedansh) [reply devsnest/app/resources/api/v1/scrum_resource.rb:15]: We can do that in controller or in resource both have before action functions

File: devsnest/db/migrate/20210622121957_adddatetoscrum.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+class Adddatetoscrum < ActiveRecord::Migration[6.0]
+  def change
+    add_column :scrums, :date, :date, :default => Date.current
```
Comment (vedansh) [inline devsnest/db/migrate/20210622121957_adddatetoscrum.rb:3]: Date.current can be wrong?, can you verify it once

File: devsnest/config/routes.rb [hunk 7]
```diff
@@ -27,6 +27,7 @@
         end
       end
       jsonapi_resources :college, only: %i[index]
+      jsonapi_resources :scrums
```
Comment (vedansh) [inline devsnest/config/routes.rb:30]: what all are allowing in scrums to happen, allow those functions

File: devsnest/app/resources/api/v1/scrum_resource.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        user_id = context[:user].id
+        if Scrum.check_scrum_auth(user_id,group)
+          super - %i[user_id group_id]
+        else
+          super - %i[attendance user_id group_id]
+        end
+      end
+
+      def self.records(options = {})
+        group_id = options[:context][:group_id] if options[:context][:user].user_type == 'admin'
```
Comment (vedansh) [inline devsnest/app/resources/api/v1/scrum_resource.rb:30]: optional, but if we are user that the context group_id is correct, we can use that one

File: devsnest/app/controllers/api/v1/scrums_controller.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        user_id = params[:data][:attributes][:user_id]
+        group = Group.where(id:params[:group_id]).first
+        unless group.group_members.where(user_id: user_id).present?
+          render_error('message': 'User is not a Part of Group')
+        end
+      end
+
+      def update_date
+        params['data']['attributes']['date'] = Date.current
+      end
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/scrums_controller.rb:39]: Ye model mai karte hai

File: devsnest/app/models/scrum.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+class Scrum < ApplicationRecord
+    validates :user_id, uniqueness: { scope: :date}
```
Comment (vedansh) [inline devsnest/app/models/scrum.rb:4]: why do we need this

Comment (sam0000-cmd) [reply devsnest/app/models/scrum.rb:4]: Yeh voh uniqueness ke liye, Docs me yehi method tha.

File: devsnest/app/controllers/api/v1/scrums_controller.rb (hunk 9 above)
Comment (sam0000-cmd) [reply devsnest/app/controllers/api/v1/scrums_controller.rb:39]: Model me before_action chalega?

File: devsnest/app/models/group.rb [hunk 11]
```diff
@@ -10,10 +10,12 @@ def parameterize
...
 
   def check_auth(user)
-    if self.group_members.where(user_id: user.id).present? || self.batch_leader_id == user.id || user.user_type == 'admin'  
-      return true
-    end
+    return true if self.group_members.where(user_id: user.id).present? || self.batch_leader_id == user.id || user.user_type == 'admin'  
+    false
+  end
 
+  def check_scrum_edit_auth(user)
```
Comment (vedansh) [inline devsnest/app/models/group.rb:17]: naame acha likh lo, jo related ho isse, something like admin

File: devsnest/app/resources/api/v1/scrum_resource.rb [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+        user = context[:user]
+        if group.check_scrum_edit_auth(user)
+          super - %i[user_id group_id date]
+        else
+          super - %i[attendance user_id group_id date]
+        end
+      end
+
+      def self.records(options = {})
+        group = Group.where(id:options[:context][:group_id]).first
```
Comment (vedansh) [inline devsnest/app/resources/api/v1/scrum_resource.rb:30]: why is this required?
